server/src/main/java/org/apache/druid/metadata/IndexerSQLMetadataStorageCoordinator.java

```diff
@@ -932,83 +930,100 @@ public int deletePendingSegments(String dataSource)
+        DataSegment segment = segmentList.get(i);
+        preparedBatch.add()
             .bind("id", segment.getId().toString())
             .bind("dataSource", segment.getDataSource())
             .bind("created_date", DateTimes.nowUtc().toString())
             .bind("start", segment.getInterval().getStart().toString())
             .bind("end", segment.getInterval().getEnd().toString())
             .bind("partitioned", (segment.getShardSpec() instanceof NoneShardSpec) ? false : true)
             .bind("version", segment.getVersion())
-            .bind("used", used)
-            .bind("payload", jsonMapper.writeValueAsBytes(segment))
-            .execute();
-
-      if (numRowsInserted == 1) {
-        log.info(
-            "Published segment [%s] to DB with used flag [%s], json[%s]",
-            segment.getId(),
-            used,
-            jsonMapper.writeValueAsString(segment)
-        );
-      } else if (numRowsInserted == 0) {
-        throw new ISE(
-            "Failed to publish segment[%s] to DB with used flag[%s], json[%s]",
-            segment.getId(),
-            used,
-            jsonMapper.writeValueAsString(segment)
-        );
-      } else {
-        throw new ISE(
-            "numRowsInserted[%s] is larger than 1 after inserting segment[%s] with used flag[%s], json[%s]",
-            numRowsInserted,
-            segment.getId(),
-            used,
-            jsonMapper.writeValueAsString(segment)
-        );
+            .bind("used", usedSegments.contains(segment))
+            .bind("payload", jsonMapper.writeValueAsBytes(segment));
+
+        if ((i + 1) % ANNOUNCE_HISTORICAL_SEGMENG_BATCH == 0 || i == segmentList.size() - 1) {
+          int[] affectedRows = preparedBatch.execute();
+          for (int j = 0; j < affectedRows.length; j++) {
+            DataSegment insertSegment = segmentList.get(i / ANNOUNCE_HISTORICAL_SEGMENG_BATCH * ANNOUNCE_HISTORICAL_SEGMENG_BATCH + j);
+            if (affectedRows[j] == 1) {
+              log.info(
+                  "Published segment [%s] to DB with used flag [%s], json[%s]",
+                  insertSegment.getId(),
+                  usedSegments.contains(insertSegment),
+                  jsonMapper.writeValueAsString(insertSegment)
+              );
+            } else {
+              throw new ISE(
+                  "Failed to publish segment[%s] to DB with used flag[%s], json[%s]",
+                  insertSegment.getId(),
+                  usedSegments.contains(insertSegment),
+                  jsonMapper.writeValueAsString(insertSegment)
+              );
+            }
+          }
+        }
       }
     }
     catch (Exception e) {
-      log.error(e, "Exception inserting segment [%s] with used flag [%s] into DB", segment.getId(), used);
+      for (DataSegment segment : segments) {
+        log.error(e, "Exception inserting segment [%s] with used flag [%s] into DB", segment.getId(), usedSegments.contains(segment));
+      }
```

Comment:
Same here. Better to be `log.errorSegments(segments, "Exception inserting segments");`